### PR TITLE
feat: dbeaver mysql compatibility, use statement and information_schema.tables

### PR DESCRIPTION
**Pull request by sunng87**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

image

This patch is to resolve compatibility issues with dbeaver client:

- it tries to access some field like `data_size` in information_schema.tables
- it uses `use` statement instead of modern `use` command which most other clients use

This patch fills these two gaps and make dbeaver client works for most cases. There is another statement to be supported: `SHOW TABLE STATUS FROM public LIKE 'system_metrics'` but it's not a blocking issue for our scenario. `SHOW TABLE STATUS` is tracked in #3354 

This patch brings back `USE` statement support in #2022 as we finally found a real use-case for it. And it also makes `current_schema` modifiable and propagate to `Session`.

The data fields of `information_schema.TABLES` in this patch are all for placeholder. I haven't fill real data into these fields. TODO item is left to future enhancement.

## Future work

I'm going refactor `Session` and `QueryContext` to use `Arc<RwLock<>>` for all modifiable fields, like time-zone, schema. This makes the change automatically propagate from `QueryContext` to `Session`. The idea of `QueryContext::update_session` is smart, but it's going to be difficult to maintain in future.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced new columns (`data_length`, `max_data_length`, `index_length`, `avg_row_length`) in the `information_schema` tables for enhanced data insight.
  - Enhanced query result structure with additional relevant columns for better data representation.

- **Improvements**
  - Implemented custom display formatting for `TableOptions` to improve readability.
  
- **Tests**
  - Added tests to verify the string conversion of `TableOptions` ensuring accurate display.
  
- **Bug Fixes**
  - Corrected column arrangement in query results to ensure consistency and correctness.

### CODE REVIEW

**Comment by killme2008** [comment]
> I'm going refactor Session and QueryContext to use Arc<RwLock<>> for all modifiable fields, like time-zone, schema. This > makes the change automatically propagate from QueryContext to Session. The idea of QueryContext::update_session is > smart, but it's going to be difficult to maintain in future.

It may hurt performance, must be careful or do some benchmark.

**Comment by killme2008** [comment]
There are many columns in `tables`, see https://dev.mysql.com/doc/refman/8.4/en/information-schema-tables-table.html

I believe we should consider including all placeholders at this time.

**Comment by sunng87** [comment]
> It may hurt performance, must be careful or do some benchmark.

Should be ok because these data are all associated with a single database connection. Actually there is no concurrent access at all. Using locks is just to bypass rust compiler check.

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by killme2008** [comment]
@evenyag @fengjiachun Please take a look.

File: src/catalog/src/information_schema/tables.rs [hunk 1]
```diff
@@ -213,9 +287,37 @@ impl InformationSchemaTablesBuilder {
...
+        self.auto_increment.push(Some(0));
+        self.row_format.push(Some("Fixed"));
+        self.table_collation.push(None);
+        self.update_time.push(None);
+        self.check_time.push(None);
+
+        self.version.push(Some(11));
+        self.table_comment.push(table_info.desc.as_deref());
+        self.create_options
+            .push(Some(format!("{:?}", table_info.meta.options).as_ref()));
```
Comment (evenyag) [inline src/catalog/src/information_schema/tables.rs:310]: Maybe serialize it to JSON format?

File: src/catalog/src/information_schema/tables.rs [hunk 2]
```diff
@@ -213,9 +287,37 @@ impl InformationSchemaTablesBuilder {
...
+        self.checksum.push(Some(0));
+        self.table_rows.push(Some(0));
+        self.data_free.push(Some(0));
+        self.auto_increment.push(Some(0));
+        self.row_format.push(Some("Fixed"));
+        self.table_collation.push(None);
+        self.update_time.push(None);
+        self.check_time.push(None);
+
+        self.version.push(Some(11));
```
Comment (evenyag) [inline src/catalog/src/information_schema/tables.rs:307]: non-blocking: Any reason to use 11 as the version?

File: src/servers/src/mysql/federated.rs [hunk 3]
```diff
@@ -249,15 +246,12 @@ fn check_show_variables(query: &str) -> Option<Output> {
...
-fn check_others(query: &str, query_ctx: QueryContextRef) -> Option<Output> {
+fn check_others(query: &str, _query_ctx: QueryContextRef) -> Option<Output> {
     if OTHER_NOT_SUPPORTED_STMT.is_match(query.as_bytes()) {
         return Some(Output::new_with_record_batches(RecordBatches::empty()));
     }
 
-    let recordbatches = if SELECT_DATABASE_PATTERN.is_match(query) {
-        let schema = query_ctx.current_schema();
-        Some(select_function("database()", schema))
-    } else if SELECT_TIME_DIFF_FUNC_PATTERN.is_match(query) {
```
Comment (evenyag) [inline src/servers/src/mysql/federated.rs:260]: Why do we remove this?

File: src/catalog/src/information_schema/tables.rs (hunk 1 above)
Comment (killme2008) [reply src/catalog/src/information_schema/tables.rs:310]: It's better in the format of `key=value` and joined with `,`, just like them in SQL.

File: src/servers/src/mysql/federated.rs (hunk 3 above)
Comment (sunng87) [reply src/servers/src/mysql/federated.rs:260]: We already have the `database()` as a built in function

File: src/catalog/src/information_schema/tables.rs (hunk 2 above)
Comment (evenyag) [reply src/catalog/src/information_schema/tables.rs:307]: Let's document this in the comment

File: src/catalog/src/information_schema/tables.rs (hunk 1 above)
Comment (sunng87) [reply src/catalog/src/information_schema/tables.rs]: The mysql way is joined with ` `, which I just updated.

File: src/catalog/src/information_schema/tables.rs (hunk 2 above)
Comment (sunng87) [reply src/catalog/src/information_schema/tables.rs:308]: I've add comments to explain this.

File: src/table/src/requests.rs [hunk 4]
```diff
@@ -128,6 +129,25 @@ impl TableOptions {
...
+            key_vals.push(format!("{}={}", TTL_KEY, humantime::Duration::from(ttl)));
+        }
+
+        for (k, v) in &self.extra_options {
+            key_vals.push(format!("{}={}", k, v));
+        }
+
+        write!(f, "{}", key_vals.join(" "))
+    }
+}
```
[the remark this answers is not in the document]
Comment (sunng87) [reply src/table/src/requests.rs:149]: I believe my original one is better

**Comment by sunng87** [comment]
@evenyag just updated, please check again

File: src/table/src/requests.rs [hunk 5]
```diff
@@ -128,6 +129,25 @@ impl TableOptions {
     }
 }
 
+impl fmt::Display for TableOptions {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut key_vals = vec![];
+        if let Some(size) = self.write_buffer_size {
+            key_vals.push(format!("{}={}", WRITE_BUFFER_SIZE_KEY, size));
```
Comment (evenyag) [inline src/table/src/requests.rs:136]: Should we quote key values by `""`?
e.g. `"write_buffer_size"="128.0MiB" "ttl"="16m 40s"`

**Review by evenyag — APPROVED** [review]

File: src/table/src/requests.rs (hunk 5 above)
Comment (killme2008) [reply src/table/src/requests.rs:136]: Looks like It's better to convert it into `OptionMap`, call its function `kv_pairs`, and join the result. This function will obscure information such as access keys.

https://github.com/GreptimeTeam/greptimedb/blob/be29e48a60345599373128c8cd6632f4f407165c/src/sql/src/statements/option_map.rs#L79

Comment (sunng87) [reply src/table/src/requests.rs:136]: It seems mysql/mariadb is writing it without quote

```
MariaDB [(none)]> select create_options from information_schema.tables limit 5;
+--------------------------------------------+
| create_options                             |
+--------------------------------------------+
| max_rows=10380835156842741 transactional=0 |
| max_rows=10720                             |
| max_rows=34952                             |
| max_rows=8587869680497929 transactional=0  |
| max_rows=51150                             |
+--------------------------------------------+
```

Comment (killme2008) [reply src/table/src/requests.rs:136]: I don't think we need to keep it compatible with MySQL or Mariadb in this column.

**Pull request merged**